src/rpc/ws_client/client.rs

```diff
@@ -15,172 +15,95 @@
 
 */
 
-use super::HandleMessage;
+use super::{subscription::WsSubscriptionWrapper, HandleMessage};
 use crate::{
-	api::{FromHexString, XtStatus},
 	rpc::{
-		json_req,
-		ws_client::{
-			GetRequestHandler, RpcClient, SubmitAndWatchHandler, SubmitOnlyHandler,
-			SubscriptionHandler,
-		},
-		Result, RpcClient as RpcClientTrait, Subscriber,
+		ws_client::{RequestHandler, RpcClient, SubscriptionHandler},
+		Request, Result, Subscribe,
 	},
+	RpcParams,
 };
-use log::info;
-use serde_json::Value;
-use sp_core::H256 as Hash;
+use serde::de::DeserializeOwned;
+use serde_json::json;
 use std::{
 	fmt::Debug,
 	sync::mpsc::{channel, Sender as ThreadOut},
 	thread,
 };
-use ws::{connect, Result as WsResult};
+use ws::{connect, Result as WsResult, Sender as WsSender};
 
 #[derive(Debug, Clone)]
 pub struct WsRpcClient {
 	url: String,
 }
 
 impl WsRpcClient {
-	pub fn new(url: &str) -> WsRpcClient {
-		WsRpcClient { url: url.to_string() }
+	pub fn new(url: &str) -> Self {
+		Self { url: url.to_string() }
 	}
-}
 
-impl RpcClientTrait for WsRpcClient {
-	fn get_request(&self, jsonreq: Value) -> Result<Option<String>> {
-		self.direct_rpc_request(jsonreq.to_string(), GetRequestHandler::default())?
-	}
-
-	fn send_extrinsic(
-		&self,
-		xthex_prefixed: String,
-		exit_on: XtStatus,
-	) -> Result<Option<sp_core::H256>> {
-		// Todo: Make all variants return a H256: #175.
-
-		let jsonreq = match exit_on {
-			XtStatus::SubmitOnly => json_req::author_submit_extrinsic(&xthex_prefixed).to_string(),
-			_ => json_req::author_submit_and_watch_extrinsic(&xthex_prefixed).to_string(),
-		};
-
-		let maybe_response =
-			self.direct_rpc_request(jsonreq, SubmitAndWatchHandler::new(exit_on))??;
-		info!("Got response {:?} while waiting for {:?}", maybe_response, exit_on);
-		match maybe_response {
-			Some(response) => Ok(Some(Hash::from_hex(response)?)),
-			None => Ok(None),
-		}
+	pub fn with_default_url() -> Self {
+		Self::new("ws://127.0.0.1:9944")
 	}
 }
 
-impl Subscriber for WsRpcClient {
-	fn start_subscriber(
-		&self,
-		json_req: String,
-		result_in: ThreadOut<<SubscriptionHandler as HandleMessage>::ThreadMessage>,
-	) -> Result<()> {
-		self.start_subscriber(json_req, result_in)
+impl Request for WsRpcClient {
+	fn request<R: DeserializeOwned>(&self, method: &str, params: RpcParams) -> Result<R> {
+		let json_req = to_json_req(method, params)?;
+		let response = self.direct_rpc_request(json_req, RequestHandler::default())??;
+		let deserialized_value: R = serde_json::from_str(&response)?;
+		Ok(deserialized_value)
 	}
 }
 
-impl WsRpcClient {
-	pub fn get(
```

Comment:
Removed a lot of boilerplate code. RpcClient should handle the rpc connection only ( subcribe and send requests), not individual functions such as `watch_until`.